rust/arrow/src/compute/kernels/substring.rs

```diff
@@ -0,0 +1,226 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines kernel to extract a substring of a StringArray
+
+use crate::{array::*, buffer::Buffer, datatypes::ToByteSlice};
+use crate::{
+    datatypes::DataType,
+    error::{ArrowError, Result},
+};
+use std::sync::Arc;
+
+/// Returns an ArrayRef with a substring starting from `start` and with optional length `length` of each of the elements in `array`.
+/// `start` can be negative, in which case the start counts from the end of the string.
+pub fn substring(array: &Array, start: i32, length: &Option<u32>) -> Result<ArrayRef> {
+    match array.data_type() {
+        DataType::Utf8 => {
+            // compute current offsets
+            let offsets = array.data_ref().clone().buffers()[0].clone();
+            let offsets: &[u32] = unsafe { offsets.typed_data::<u32>() };
+
+            // compute null bitmap (copy)
+            let null_bit_buffer = array.data_ref().null_buffer().map(|e| e.clone());
+
+            // compute values
+            let values = &array.data_ref().buffers()[1];
+            let data = values.data();
+
+            let mut new_values = Vec::new(); // we have no way to estimate how much this will be.
```

Comment:
I think we can estimate the capacity as `length * array.len()` if the length is specified and `values.len() - (start*array.len())` otherwise.

Hm, actually depends on the semantics of the `length` parameter, is it supposed to be a maximum length or should it be an error if `start+length` goes out of bounds of one entry?

Reply:
I think that this all depends on what assumptions we make about a typical query. Specifically:

* what is the average start
* what is the average length
* what is the average string size

from these 3, we could compute an initial estimation. For example: is this going to be used to treat texts, on which "string size" can go to Mbs, or is this going to be used on 10-12 character arrays?

I considered this discussion beyond the scope of this PR and just went with the least memory-intensive/highest CPU-intensive approach, just because 100% memory usage usually has more nefarious implications than 100% CPU.

The semantics is so that it does not error: it just goes to the maximum/minimum available for each entry. The rational here is that otherwise the user would have to perform some calculation on the side to ensure no errors at runtime, which are never fun. :)